NTupleMaker/interface/functionsCP.h

```diff
@@ -34,215 +35,184 @@ TLorentzVector gen_ipVec(const AC1B * analysisTree, int tauIndex, int piIndex, T
 int gen_chargedPiIndex(const AC1B * analysisTree, int tauIndex, int partId);
 
 TLorentzVector ipVec_Lepton(const AC1B * analysisTree, int tauIndex, TString ch);
-  
-double acoCP(TLorentzVector Pi1, TLorentzVector Pi2, 
-	     TLorentzVector ref1, TLorentzVector ref2,
-	     bool firstNegative, bool pi01, bool pi02) {
-
-  double y1 = 1;
-  double y2 = 1;
-  if (pi01)
-    y1 = Pi1.E() - ref1.E();
-  if (pi02)
-    y2 = Pi2.E() - ref2.E();
-
-
-  double y = y1*y2; 
 
-  TLorentzVector Prongsum = Pi1 + Pi2;
-  TVector3 boost = -Prongsum.BoostVector();
-  Pi1.Boost(boost);
-  Pi2.Boost(boost);
-  ref1.Boost(boost);
-  ref2.Boost(boost);
-
-  //  std::cout << "First negative = " << firstNegative << "  pi01 = " << pi01 << "   pi02 = " << pi02 << std::endl;
-  //  std::cout << "Px(1) = " << Pi1.Px() << "  Py(1) = " << Pi1.Py() << "  Pz(1) = " << Pi1.Pz() << std::endl;
-  //  std::cout << "Px(2) = " << Pi2.Px() << "  Py(2) = " << Pi2.Py() << "  Pz(2) = " << Pi2.Pz() << std::endl;
-  //  std::cout << "Ux(1) = " << ref1.Px() << "  Uy(1) = " << ref1.Py() << "  Uz(1) = " << ref1.Pz() << std::endl;
-  //  std::cout << "Ux(2) = " << ref2.Px() << "  Uy(2) = " << ref2.Py() << "  Uz(2) = " << ref2.Pz() << std::endl;
-  //  std::cout << std::endl;
-  // get 3-vectors
-  TVector3 vecPi1 = Pi1.Vect();
-  TVector3 vecPi2 = Pi2.Vect();
-  TVector3 vecRef1 = ref1.Vect();
-  TVector3 vecRef2 = ref2.Vect();
-
-  // normalize them 
-  vecPi1 *= 1/vecPi1.Mag();
-  vecPi2 *= 1/vecPi2.Mag();
-  vecRef1 *= 1/vecRef1.Mag();
-  vecRef2 *= 1/vecRef2.Mag();
-
-  // transverse components  
-  TVector3 vecRef1transv = vecRef1 - vecPi1*(vecPi1*vecRef1);
-  TVector3 vecRef2transv = vecRef2 - vecPi2*(vecPi2*vecRef2);
-
-  vecRef1transv *= 1/vecRef1transv.Mag();
-  vecRef2transv *= 1/vecRef2transv.Mag();
-
-  double acop = TMath::ACos(vecRef1transv*vecRef2transv);
-  
-  double sign = vecPi2 * vecRef1transv.Cross(vecRef2transv);
-  if (firstNegative)
-    sign = vecPi1 * vecRef2transv.Cross(vecRef1transv);
-  if (sign<0) acop = 2.0*TMath::Pi() - acop;
-
-  if (y<0) {
-    acop = acop + TMath::Pi();
-    if (acop>2*TMath::Pi()) {
-      acop = acop - 2*TMath::Pi();
-    }
-  }  
-
-//cout<<"acop before return from function "<<acop<<endl;
-  return acop;
-
-}
-
-//Merijn added these functions for debugging purposes
-double acoCP2(TLorentzVector Pi1, TLorentzVector Pi2, 
+double acoCP(TLorentzVector Pi1, TLorentzVector Pi2, 
 	     TLorentzVector ref1, TLorentzVector ref2,
-	     bool firstNegative, bool pi01, bool pi02);
-
-
+	     bool firstNegative, bool pi01, bool pi02, Synch17Tree *otree);
```

Comment:
Why do you need the pointer to the Synch17Tree in a function that was meant to work only with Lorentz Vectors?
I see that you placed here the initialization of the variables in the SynchNTuples, but given that the function just returns a double it would make more sense to have the initialization either in acott() where the tree pointers are accessed either way, or in the functions in the SynchNTupleProducer...

Reply:
Hi Andrea, beside of saving the aco angle phi, I started to save also psi to do some cross checks if the distribution looks sane, also for gen level. This is calculated in this function.
Currently, as a quick method, I overloaded the function.
We could decide also to return a double * with phi and psi, and assign the angles to the tree when the functions has finished. Perhaps for now keep, and do something more sophisticated later?

Reply:
Yes, seems fine to me